modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/CheckpointTest.java

```diff
@@ -0,0 +1,273 @@
+                        assertEquals(lastWriteEntryDuration.value(), writeCpEntryDuration);
+                        assertEquals(lastSplitAndSortPagesDuration.value(), splitAndSortCpPagesDuration);
+                        assertEquals(lastDuration.value(), totalDuration);
+                        assertEquals(lastStart.value(), cpStartTime);
+                        assertEquals(lastTotalPages.value(), pagesSize);
+                        assertEquals(lastDataPages.value(), dataPagesWritten);
+                        assertEquals(lastCOWPages.value(), cowPagesWritten);
+
+                        cnt.incrementAndGet();
+                    }
+                });
+
+                return cnt.get() == 1;
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }, TIMEOUT));
+    }
+
+    /** @throws Exception if failed. */
+    @Test
+    public void testThrottleSpeedBased() throws Exception {
+        checkThrottling();
+    }
+
+    /** @throws Exception if failed. */
+    @Test
+    @WithSystemProperty(key = IGNITE_OVERRIDE_WRITE_THROTTLING_ENABLED, value = "TARGET_RATIO_BASED")
+    public void testThrottleTargetRatioBased() throws Exception {
+        checkThrottling();
+    }
+
+    /** @throws Exception if failed. */
+    public void checkThrottling() throws Exception {
+        IgniteEx srv = startGrid();
+
+        srv.cluster().state(ClusterState.ACTIVE);
+
+        IgniteCache<Long, Long> cache = srv.getOrCreateCache(DEFAULT_CACHE_NAME);
+
+        long start = U.currentTimeMillis();
+
+        startCollectStatisticsWithImmediatelyFlush();
+
+        int keysCnt = 1024;
+
+        AtomicBoolean stop = new AtomicBoolean();
+
+        slowCheckpointEnabled.set(true);
+
+        IgniteInternalFuture<?> fut = GridTestUtils.runAsync(() -> {
+            while (!stop.get())
+                cache.put(ThreadLocalRandom.current().nextLong(keysCnt), ThreadLocalRandom.current().nextLong());
+        });
+
+        assertTrue(waitForCondition(() -> {
+            try {
+                AtomicInteger cnt = new AtomicInteger();
+
+                readFiles(statisticsFiles(), new TestHandler() {
+                    @Override public void pagesWriteThrottle(UUID nodeId, long endTime, long duration) {
+                        assertEquals(srv.localNode().id(), nodeId);
+
+                        assertTrue(start <= endTime);
+                        assertTrue(duration >= 0);
+
+                        cnt.incrementAndGet();
+                    }
+                });
+
+                return cnt.get() > 0;
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }, TIMEOUT));
+
+        stop.set(true);
+
+        fut.get(TIMEOUT);
+    }
+
+    /**
+     * Create File I/O that emulates poor checkpoint write speed.
+     */
+    private static class SlowCheckpointFileIOFactory implements FileIOFactory {
```

Comment:
Let's reuse the same class from `PagesWriteThrottleSmokeTest` or `CheckpointBufferDeadlockTest` instead of copy-pasting it.

Reply:
fixed